uint8 type
uint8 rolled
uint8 threshold